src/RequestProcessor.cs

```diff
@@ -299,27 +321,15 @@ internal StreamingMessage ProcessFunctionEnvironmentReloadRequest(StreamingMessa
         private void InitializeForFunctionApp(StreamingMessage request, StreamingMessage response)
         {
             var functionLoadRequest = request.FunctionLoadRequest;
-
-            // If 'ManagedDependencyEnabled' is true, process the function app dependencies as defined in FunctionAppRoot\requirements.psd1.
-            // These dependencies will be installed via 'Save-Module' when the first PowerShellManager instance is being created.
-            if (functionLoadRequest.ManagedDependencyEnabled)
-            {
-                _dependencyManager.Initialize(functionLoadRequest);
-            }
-
             // Setup the FunctionApp root path and module path.
-            FunctionLoader.SetupWellKnownPaths(functionLoadRequest);
-
-            // Constructing the first PowerShellManager instance for the Pool.
-            if (DependencyManager.Dependencies.Count > 0)
-            {
-                // Do extra work to install the specified dependencies.
-                _powershellPool.Initialize(request.RequestId, _dependencyManager.InstallFunctionAppDependencies);
-                response.FunctionLoadResponse.IsDependencyDownloaded = true;
-            }
-            else
+            FunctionLoader.SetupWellKnownPaths(request.FunctionLoadRequest);
+            if (functionLoadRequest.ManagedDependencyEnabled)
             {
-                _powershellPool.Initialize(request.RequestId);
+                //Start dependency download on a separate thread
+                _dependencyDownloadTask = Task.Run(() => _dependencyManager.ProcessDependencies(_msgStream, request)).ContinueWith((task) =>
```

Comment:
The task creation and the reference to the task should be move to `DependencyManager`. For example, `DependencyManager.Initialize(...)` creates the task and hold a reference to it within `DependencyManager`. In `ProcessInvocationRequest` we can wait on the task by `_dependencyManager.DownloadTask.Wait()` or `_dependencyManager.WaitDownloadToFinish()`.

Reply:
I think better approach would be to return a Rpc messag and block until dependencies are downloaded as invocation will not succeed before download completes.

Reply:
We should add another state to the [StatusResult](https://github.com/Azure/azure-functions-language-worker-protobuf/blob/dev/src/proto/FunctionRpc.proto#L112) `Loading`  to allow host to handle intermediate state where function is still loading but not ready for invocation

Reply:
That sounds like a much better long term solution

Reply:
The PM requirement is to show the message on the azure portal when a function is invoked through portal while managed dependency download is in-progress